From 86670553fa505b0b243ba6eb31e7f3f80b7e21b2 Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Fri, 26 May 2017 14:11:21 +0800
Subject: [PATCH 1022/1345] fix: usb: gadget: u3d: remove HW reset from probe
 for a3700

commit  287f999f538d2710c05c3fd8abfafbea7f3748c7 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

According to Marvell USB3 device IP specification,
the HW reset is not necessary in U3D probe.

The patch removed HW reset from U3D probe routine only
for SoC of armada3700.

Change-Id: Idaf44f190bad0214623984d4caf490f8572308a0
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/40217
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/gadget/udc/mvebu_u3d.c |   10 +++++++++-
 drivers/usb/gadget/udc/mvebu_u3d.h |    2 ++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/gadget/udc/mvebu_u3d.c b/drivers/usb/gadget/udc/mvebu_u3d.c
index 6cae655..de66bf0 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.c
+++ b/drivers/usb/gadget/udc/mvebu_u3d.c
@@ -2429,6 +2429,13 @@ static int mvc2_probe(struct platform_device *pdev)
 		cp->reg->global_control = 0x24;
 	}
 
+	/* For Armada 3700, need to skip PHY HW reset */
+	if (of_device_is_compatible(pdev->dev.of_node,
+				    "marvell,armada3700-u3d"))
+		cp->phy_hw_reset = false;
+	else
+		cp->phy_hw_reset = true;
+
 	/* Get comphy and init if there is */
 	cp->comphy = devm_of_phy_get(&pdev->dev, pdev->dev.of_node, "usb");
 	if (!IS_ERR(cp->comphy)) {
@@ -2460,7 +2467,8 @@ static int mvc2_probe(struct platform_device *pdev)
 
 	eps_init(cp);
 
-	mvc2_hw_reset(cp);
+	if (cp->phy_hw_reset)
+		mvc2_hw_reset(cp);
 
 	dev_set_drvdata(cp->dev, cp);
 	dev_info(cp->dev, "Detected ver %x from Marvell Central IP.\n", ver);
diff --git a/drivers/usb/gadget/udc/mvebu_u3d.h b/drivers/usb/gadget/udc/mvebu_u3d.h
index 330125f..b49e900 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.h
+++ b/drivers/usb/gadget/udc/mvebu_u3d.h
@@ -394,6 +394,8 @@ struct mvc2 {
 	int prev_vbus;
 	struct work_struct	vbus_work;
 	struct workqueue_struct *qwork;
+	/* Flags for HW reset. false: no need reset; true: need reset */
+	bool phy_hw_reset;
 };
 
 extern void mvc2_usb2_connect(void);
-- 
1.7.9.5

